sns/parity_math.c

```diff
@@ -500,29 +551,110 @@ static void reed_solomon_encode(struct m0_parity_math *math,
 
 #undef PARITY_MATH_REGION_ENABLE
 }
+#endif /* __KERNEL__ */
+
+#ifndef __KERNEL__
+static void isal_encode(struct m0_parity_math *math,
+			const struct m0_buf *data,
+			struct m0_buf *parity)
+{
+	uint8_t **data_frags = NULL;
+	uint8_t **parity_frags = NULL;
+	uint32_t  i;
+	uint32_t  data_count;
+	uint32_t  parity_count;
+	uint32_t  block_size;
+
+	M0_PRE(math != NULL);
+	M0_PRE(data != NULL);
+	M0_PRE(parity != NULL);
+
+	M0_ENTRY();
+
+	data_count = math->pmi_data_count;
+	parity_count = math->pmi_parity_count;
+	block_size = data[0].b_nob;
+
+	M0_ALLOC_ARR(data_frags, data_count);
+	if (data_frags == NULL) {
+		M0_ERR_INFO(-ENOMEM, "failed to allocate memory for array of "
+			    "data fragments");
+		goto fini;
+	}
+
+	M0_ALLOC_ARR(parity_frags, parity_count);
+	if (parity_frags == NULL) {
+		M0_ERR_INFO(-ENOMEM, "failed to allocate memory for array of "
+			    "parity fragments");
+		goto fini;
+	}
+
+	data_frags[0] = (uint8_t *)data[0].b_addr;
+	for (i = 1; i < data_count; ++i) {
+		if (block_size != data[i].b_nob) {
+			M0_ERR_INFO(-errno, "data block size mismatch. "
```

Comment:
Please don't use the external variable 'errno'.
At this point, the 'external' may contain irrelevant value.
Just return -EINVAL.